src/System.Net.Http/src/System/Net/Http/Managed/HttpProxyConnectionHandler.cs

```diff
@@ -0,0 +1,146 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Concurrent;
+using System.IO;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http
+{
+    internal sealed class HttpProxyConnectionHandler : HttpMessageHandler
+    {
+        private readonly IWebProxy _proxy;
+        private readonly HttpMessageHandler _innerHandler;
+
+        private readonly ConcurrentDictionary<HttpConnectionKey, HttpConnectionPool> _connectionPoolTable;
+        private bool _disposed;
+
+        public HttpProxyConnectionHandler(IWebProxy proxy, HttpMessageHandler innerHandler)
+        {
+            if (proxy == null)
+            {
+                throw new ArgumentNullException(nameof(proxy));
+            }
+
+            if (innerHandler == null)
+            {
+                throw new ArgumentNullException(nameof(innerHandler));
+            }
+
+            _proxy = proxy;
+            _innerHandler = innerHandler;
+
+            _connectionPoolTable = new ConcurrentDictionary<HttpConnectionKey, HttpConnectionPool>();
+        }
+
+        protected internal override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            Uri proxyUri = null;
+            try
+            {
+                if (!_proxy.IsBypassed(request.RequestUri))
+                {
+                    proxyUri = _proxy.GetProxy(request.RequestUri);
+                }
+            }
+            catch (Exception)
+            {
+                // Eat any exception from the IWebProxy and just treat it as no proxy.
+                // TODO: This seems a bit questionable, but it's what the tests expect
+            }
+
+            if (proxyUri == null)
+            {
+                return await _innerHandler.SendAsync(request, cancellationToken);
+            }
+
+            if (proxyUri.Scheme != "http")
+            {
+                throw new InvalidOperationException($"invalid scheme {proxyUri.Scheme} for proxy");
+            }
+
+            if (request.RequestUri.Scheme == "https")
+            {
+                // TODO: Implement SSL tunneling through proxy
+                throw new NotImplementedException("no support for SSL tunneling through proxy");
+            }
+
+            HttpConnection connection = await GetOrCreateConnection(request, proxyUri);
+
+            HttpResponseMessage response = await connection.SendAsync(request, cancellationToken);
+
+            // Handle proxy authentication
+            if (response.StatusCode == HttpStatusCode.ProxyAuthenticationRequired &&
+                _proxy.Credentials != null)
+            {
+                foreach (AuthenticationHeaderValue h in response.Headers.ProxyAuthenticate)
+                {
+                    // We only support Basic auth, ignore others
```

Comment:
I have the bones of a Digest auth implementation in the WCF codebase (I have the service side implemented, client shouldn't be much different) I'll try and find time to rework for this in the coming weeks. I did prototype using hashes other than md5 on the service side and none of our existing implementations supported anything other than md5 so I'm looking forward to having a digest auth implementation which is up to date with the latest rfc's.